# Introduction to SQL


## Relational Databases



In [ ]:
pip install sqlalchemy psycopg2 ipython-sql

In [2]:
%load_ext sql


In [5]:
%sql postgresql://postgres:postgres@localhost/local


In [6]:
%config SqlMagic.autolimit = 10


### Our very own table

A database has been set up for this course and the `books` table is
available here.

Run the code to explore what data `books` holds!

**Instructions**

- Hit "Run Code" to see the `books` table.

**Answer**


In [7]:
%%sql
SELECT *
FROM books;


 * postgresql://postgres:***@localhost/local
350 rows affected.


[(1, '10-Day Green Smoothie Cleanse', 'JJ Smith', 2016, 'Non Fiction'),
 (2, '11/22/63: A Novel', 'Stephen King', 2011, 'Fiction'),
 (3, '12 Rules for Life: An Antidote to Chaos', 'Jordan B. Peterson', 2018, 'Non Fiction'),
 (4, '1984 (Signet Classics)', 'George Orwell', 2017, 'Fiction'),
 (5, '"5,000 Awesome Facts (About Everything!) (National Geographic Kids)"', 'National Geographic Kids', 2019, 'Non Fiction'),
 (6, 'A Dance with Dragons (A Song of Ice and Fire)', 'George R. R. Martin', 2011, 'Fiction'),
 (7, 'A Game of Thrones / A Clash of Kings / A Storm of Swords / A Feast of Crows / A Dance with Dragons', 'George R. R. Martin', 2014, 'Fiction'),
 (8, 'A Gentleman in Moscow: A Novel', 'Amor Towles', 2017, 'Fiction'),
 (9, '"A Higher Loyalty: Truth, Lies, and Leadership"', 'James Comey', 2018, 'Non Fiction'),
 (10, 'A Man Called Ove: A Novel', 'Fredrik Backman', 2016, 'Fiction')]

## Querying

### Querying the books table

You're ready to practice writing your first SQL queries using the
`SELECT` and `FROM` keywords. Recall from the video that `SELECT` is
used to choose the fields that will be included in the result set, while
`FROM` is used to pick the table in which the fields are listed.

Feel free to explore `books` in the exercise.

Your task in this exercise is to practice selecting fields from `books`.

**Instructions**

- Use SQL to return a result set of all book titles included in the
  `books` table.

<!-- -->

- Select both the `title` and `author` fields from `books`.

<!-- -->

- Select all fields from the `books` table.

**Answer**


In [8]:
%%sql
-- Return all titles from the books table
SELECT title
FROM books;

-- Select title and author from the books table
SELECT title, author
FROM books;

-- Select all fields from the books table
SELECT *
FROM books;


 * postgresql://postgres:***@localhost/local
350 rows affected.
350 rows affected.
350 rows affected.


[(1, '10-Day Green Smoothie Cleanse', 'JJ Smith', 2016, 'Non Fiction'),
 (2, '11/22/63: A Novel', 'Stephen King', 2011, 'Fiction'),
 (3, '12 Rules for Life: An Antidote to Chaos', 'Jordan B. Peterson', 2018, 'Non Fiction'),
 (4, '1984 (Signet Classics)', 'George Orwell', 2017, 'Fiction'),
 (5, '"5,000 Awesome Facts (About Everything!) (National Geographic Kids)"', 'National Geographic Kids', 2019, 'Non Fiction'),
 (6, 'A Dance with Dragons (A Song of Ice and Fire)', 'George R. R. Martin', 2011, 'Fiction'),
 (7, 'A Game of Thrones / A Clash of Kings / A Storm of Swords / A Feast of Crows / A Dance with Dragons', 'George R. R. Martin', 2014, 'Fiction'),
 (8, 'A Gentleman in Moscow: A Novel', 'Amor Towles', 2017, 'Fiction'),
 (9, '"A Higher Loyalty: Truth, Lies, and Leadership"', 'James Comey', 2018, 'Non Fiction'),
 (10, 'A Man Called Ove: A Novel', 'Fredrik Backman', 2016, 'Fiction')]

### Making queries DISTINCT

You've learned that the `DISTINCT` keyword can be used to return unique
values in a field. In this exercise, you'll use this understanding to
find out more about the `books` table!

There are 350 books in the `books` table, representing all of the books
that our local library has available for checkout. But how many
different authors are represented in these 350 books? The answer is
surely less than 350. For example, J.K. Rowling wrote all seven Harry
Potter books, so if our library has all Harry Potter books, seven books
will be written by J.K Rowling. There are likely many more repeat
authors!

**Instructions**

- Write SQL code that returns a result set with just one column listing
  the unique authors in the `books` table.

<!-- -->

- Update the code to return the unique `author` and `genre` combinations
  in the `books` table.

**Answer**


In [9]:
%%sql
-- Select unique authors from the books table
SELECT DISTINCT author
FROM books;

-- Select unique authors and genre combinations from the books table
SELECT DISTINCT author, genre
FROM books;


 * postgresql://postgres:***@localhost/local
247 rows affected.
249 rows affected.


[('Eben Alexander', 'Non Fiction'),
 ('Adam Mansbach', 'Fiction'),
 ('Garth Stein', 'Fiction'),
 ('Gillian Flynn', 'Fiction'),
 ('Tucker Carlson', 'Non Fiction'),
 ('Atul Gawande', 'Non Fiction'),
 ('Wizards RPG Team', 'Fiction'),
 ('Giles Andreae', 'Fiction'),
 ('Dav Pilkey', 'Fiction'),
 ('Malcolm Gladwell', 'Non Fiction')]

### Aliasing

While the default column names in a SQL result set come from the fields
they are created from, you've learned that aliasing can be used to
rename these result set columns. This can be helpful for clarifying the
intent or contents of the column.

Your task in this exercise is to incorporate an alias into one of the
SQL queries that you worked with in the previous exercise!

**Instructions**

- Add an alias to the SQL query to rename the `author` column to
  `unique_author` in the result set.

**Answer**


In [10]:
%%sql
-- Alias author so that it becomes unique_author
SELECT DISTINCT author AS unique_author
FROM books;


 * postgresql://postgres:***@localhost/local
247 rows affected.


[('John Heilemann',),
 ('Sheryl Sandberg',),
 ('Brené Brown',),
 ('DK',),
 ('Rebecca Skloot',),
 ('Crispin Boyer',),
 ('Daniel James Brown',),
 ('Bruce Springsteen',),
 ('Fredrik Backman',),
 ('Heidi Murkoff',)]

### VIEWing your query

You've worked hard to create the below SQL query:

    SELECT DISTINCT author AS unique_author
    FROM books;

What if you'd like to be able to refer to it later, or allow others to
access and use the results? The best way to do this is by creating a
view. Recall that a view is a virtual table: it's very similar to a real
table, but rather than the data itself being stored, the query code is
stored for later use.

**Instructions**

- Add a single line of code that saves the results of the written query
  as a view called `library_authors`.

**Answer**


In [11]:
%%sql
-- Save the results of this query as a view called library_authors
CREATE VIEW library_authors AS
SELECT DISTINCT author AS unique_author
FROM books;

-- Your code to create the view:
CREATE VIEW library_authors AS
SELECT DISTINCT author AS unique_author
FROM books;

-- Select all columns from library_authors
SELECT *
FROM library_authors;


 * postgresql://postgres:***@localhost/local
Done.
(psycopg2.errors.DuplicateTable) relation "library_authors" already exists

[SQL: -- Your code to create the view:
CREATE VIEW library_authors AS
SELECT DISTINCT author AS unique_author
FROM books;]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Limiting results

Let's take a look at a few of the genres represented in our library's
books.

Recall that limiting results is useful when testing code since result
sets can have thousands of results! Queries are often written with a
`LIMIT` of just a few records to test out code before selecting
thousands of results from the database.

Let's practice with `LIMIT`!

**Instructions**

- Using PostgreSQL, select the `genre` field from the `books` table;
  limit the number of results to 10.

**Answer**


In [12]:
%%sql
-- Select the first 10 genres from books using PostgreSQL
SELECT genre
FROM books
LIMIT 10;


 * postgresql://postgres:***@localhost/local
10 rows affected.


[('Non Fiction',),
 ('Fiction',),
 ('Non Fiction',),
 ('Fiction',),
 ('Non Fiction',),
 ('Fiction',),
 ('Fiction',),
 ('Fiction',),
 ('Non Fiction',),
 ('Fiction',)]